### PR TITLE
Bridge NSNull to CFNullRef and [NSNull null] to kCFNull

### LINKED ISSUE #2547: NSNull is not bridged to CFNullRef, breaking subtle CF things
Some parts of CoreFoundation ([CFURLComponents](https://github.com/Microsoft/WinObjC/blob/develop/Frameworks/CoreFoundation/URL.subproj/CFURLComponents.c#L1204), [CFSocket](https://github.com/Microsoft/WinObjC/blob/develop/Frameworks/CoreFoundation/RunLoop.subproj/CFSocket.c#L2302)) use direct pointer equality against `kCFNull`, the `CFNullRef` singleton, to determine object residence.

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
Fixes #2547

### CODE REVIEW

File: Frameworks/Foundation/NSNull.mm [hunk 1]
```diff
@@ -38,25 +58,24 @@ - (instancetype)initWithCoder:(NSCoder*)coder {
...
 - (void)encodeWithCoder:(NSCoder*)coder {
-    UNIMPLEMENTED();
 }
 
 /**
  @Status Interoperable
 */
 - (instancetype)copyWithZone:(NSZone*)zone {
-    return nullSingleton;
+    return static_cast<NSNull*>(kCFNull);
```
Comment (aballway) [inline Frameworks/Foundation/NSNull.mm]: nit: could make these all just return [NSNull null]

File: tests/unittests/Foundation/NSNullTests.mm [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    const unsigned char rawPlist[] = { 'b',  'p',  'l',  'i',  's',  't',  '0',  '0',  0x00, 0x08, 0x00, 0x00, 0x00, 0x00,
+                                       0x00, 0x00, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00,
+                                       0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x09 };
+    NSData* plistData = [NSData dataWithBytes:rawPlist length:std::extent<decltype(rawPlist)>::value];
+    NSObject* nullObject = nil;
+    ASSERT_NO_THROW(
+        nullObject =
+            [NSPropertyListSerialization propertyListFromData:plistData mutabilityOption:0 format:nullptr errorDescription:nullptr]);
+    ASSERT_NE(nil, nullObject);
```
Comment (aballway) [inline tests/unittests/Foundation/NSNullTests.mm]: nit: unneeded, following test will catch this (unless [NSNull null] returns NULL in which case we should have a test that verifies NE([NSNull null], NULL)

Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSNullTests.mm]: there is such a test ;P

File: Frameworks/Foundation/NSNull.mm (hunk 1 above)
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSNull.mm]: eh, one more message dispatch

File: tests/unittests/Foundation/NSNullTests.mm [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+#import <TestFramework.h>
+#import <Foundation/Foundation.h>
+
+TEST(NSNull, Bridged) {
+    ASSERT_EQ([NSNull null], kCFNull); // Pointer equality required.
+}
+
+TEST(NSNull, Singleton) {
+    ASSERT_NE(nil, [NSNull null]);
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSNullTests.mm]: is good

**Comment by DHowett-MSFT** [comment]
@ms-jihua likely, but I'd like to keep that cleanup out of this PR

**Pull request merged**